### PR TITLE
executor, metrics: add metrics for expensive executors and statement nodes

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: executor/compiler.go [hunk 1]
```diff
@@ -46,17 +49,151 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
 		Ctx:        c.Ctx,
 	}, nil
 }
 
+func logExpensiveQuery(stmtNode ast.StmtNode) {
+	const logSQLLen = 1024
+
+	sql := stmtNode.Text()
+	if len(sql) > logSQLLen {
+		sql = fmt.Sprintf("%s len(%d)", sql[:logSQLLen], len(sql))
```
Comment (XuHuaiyu) [inline executor/compiler.go]:
``` go
str := "abc张三"
fmt.Println(str[:4])
```

will not get `abc张`, but `abc?`.

File: executor/compiler.go [hunk 2]
```diff
@@ -46,17 +49,151 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
+	switch x := p.(type) {
+	case plan.PhysicalPlan:
+		return isPhysicalPlanExpensive(x)
+	default:
+		return false
+	}
+}
+
+func isPhysicalPlanExpensive(p plan.PhysicalPlan) bool {
+	if p.StatsInfo().Count() > 10000 {
```
Comment (XuHuaiyu) [inline executor/compiler.go]: make 10000 a constant?

File: executor/compiler.go [hunk 3]
```diff
@@ -46,17 +49,151 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
+	const logSQLLen = 1024
+
+	sql := stmtNode.Text()
+	if len(sql) > logSQLLen {
+		sql = fmt.Sprintf("%s len(%d)", sql[:logSQLLen], len(sql))
+	}
+	log.Warnf("[EXPENSIVE_QUERY] %s", sql)
+}
+
+func isExpensive(p plan.Plan) bool {
```
Comment (XuHuaiyu) [inline executor/compiler.go]: s/ isExpensive/ isExpensivePlan or isExpensiveQuery ?

Comment (XuHuaiyu) [reply executor/compiler.go]: why not define `isExpensive(p plan.Plan)` as `isExpensive(p plan.PhysicalPlan)`

File: executor/compiler.go [hunk 4]
```diff
@@ -46,17 +49,151 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
 		return nil, errors.Trace(err)
 	}
 
+	CountStmtNode(stmtNode, c.Ctx.GetSessionVars().InRestrictedSQL)
+	isExpensive := isExpensive(finalPlan)
```
Comment (XuHuaiyu) [inline executor/compiler.go]: put this check into `logExpensiveQuery`?

File: executor/compiler.go (hunk 3 above)
Comment (XuHuaiyu) [reply executor/compiler.go]: Will this func log sql like `delete from hashjoin`?

File: executor/compiler.go (hunk 1 above)
Comment (zz-jason) [reply executor/compiler.go]: yes, it does not matter how long the text is since it's already been truncated.

**Comment by zz-jason** [comment]
@tiancaiamao @XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
CI failed. @zz-jason

File: executor/compiler.go [hunk 5]
```diff
@@ -46,17 +49,168 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
+}
+
+func isExpensiveQuery(p plan.Plan) bool {
+	switch x := p.(type) {
+	case plan.PhysicalPlan:
+		return isPhysicalPlanExpensive(x)
+	case *plan.Execute:
+		return isExpensiveQuery(x.Plan)
+	case *plan.Insert:
+		if x.SelectPlan != nil {
```
Comment (XuHuaiyu) [inline executor/compiler.go]: Seems this check is not necessary?

File: executor/builder.go [hunk 6]
```diff
@@ -1283,6 +1292,7 @@ func (b *executorBuilder) buildIndexLookUpReader(v *plan.PhysicalIndexLookUpRead
 
 	is := v.IndexPlans[0].(*plan.PhysicalIndexScan)
 	ret.ranges = is.Ranges
+	metrics.ExecutorCounter.WithLabelValues("IndexLookUpExecutor").Inc()
```
Comment (XuHuaiyu) [inline executor/builder.go]: IndexLookUpReader

File: executor/executor_test.go [hunk 7]
```diff
@@ -2090,10 +2090,10 @@ func (s *testContextOptionSuite) TestCoprocessorPriority(c *C) {
 	cli.priority = pb.CommandPri_Low
 	tk.MustQuery("select count(*) from t")
 
-	cli.priority = pb.CommandPri_Low
+	cli.priority = pb.CommandPri_Normal
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: why we need change this?

File: executor/builder.go (hunk 6 above)
Comment (zz-jason) [reply executor/builder.go]: the executor is `IndexLookUpExecutor`

File: executor/compiler.go (hunk 5 above)
Comment (zz-jason) [reply executor/compiler.go]: no

File: executor/executor_test.go (hunk 7 above)
Comment (zz-jason) [reply executor/executor_test.go]: the logical to define whether a query is expensive has changed.

File: executor/compiler.go [hunk 8]
```diff
@@ -46,17 +49,168 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
+			return true
+		}
+	}
+
+	switch x := p.(type) {
+	case *plan.PhysicalTopN:
+		return true
+	case *plan.PhysicalSort:
+		return true
+	case *plan.NominalSort:
```
Comment (lamxTyler) [inline executor/compiler.go]: Why is a nominal sort expensive?

Comment (zz-jason) [reply executor/compiler.go]: "NominalSort" should not appear in the final physical plan, removed.

File: executor/compiler.go [hunk 9]
```diff
@@ -46,17 +49,166 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
+	if !expensive {
+		return
+	}
+
+	const logSQLLen = 1024
+	sql := stmtNode.Text()
+	if len(sql) > logSQLLen {
+		sql = fmt.Sprintf("%s len(%d)", sql[:logSQLLen], len(sql))
+	}
+	log.Warnf("[EXPENSIVE_QUERY] %s", sql)
```
Comment (tiancaiamao) [inline executor/compiler.go]: I'm afraid there would too many EXPENSIVE_QUERY after this change.

File: executor/executor_test.go (hunk 7 above)
Comment (tiancaiamao) [reply executor/executor_test.go]: Maybe we should insert more data in this test, to make it cover that full table scan is expensive thus have low priority.

File: executor/compiler.go (hunk 9 above)
Comment (zz-jason) [reply executor/compiler.go]: any suggestion ?

File: executor/compiler.go [hunk 10]
```diff
@@ -46,17 +49,144 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
...
+	case *ast.DropDatabaseStmt:
+		return "DropDatabase"
+	case *ast.DropIndexStmt:
+		return "DropIndex"
+	case *ast.DropTableStmt:
+		return "DropTable"
+	case *ast.ExplainStmt:
+		return "Explain"
+	case *ast.InsertStmt:
+		if x.IsReplace {
```
Comment (tiancaiamao) [inline executor/compiler.go]: I'm not sure this works, it seems Replace never show in our prometheus?

File: executor/executor_test.go [hunk 11]
```diff
@@ -2083,13 +2083,13 @@ func (s *testContextOptionSuite) TestCoprocessorPriority(c *C) {
 	tk.MustQuery("select id from t where id = 1")
 	tk.MustQuery("select * from t1 where id = 1")
 
-	cli.priority = pb.CommandPri_Low
+	cli.priority = pb.CommandPri_Normal
```
Comment (tiancaiamao) [inline executor/executor_test.go]: I still think we should make something to cover the low priority case...

File: executor/compiler.go (hunk 10 above)
Comment (zz-jason) [reply executor/compiler.go]: It works, I've added a test case in **server/server_test.go**

File: executor/executor_test.go [hunk 12]
```diff
@@ -2012,6 +2013,7 @@ func (c *checkRequestClient) SendReq(ctx goctx.Context, addr string, req *tikvrp
 		switch req.Type {
 		case tikvrpc.CmdCop:
 			if c.priority != req.Priority {
+				fmt.Printf("expected proority: %v, actual priority: %v\n", c.priority, req.Priority)
```
Comment (tiancaiamao) [inline executor/executor_test.go]: remove debug log

File: executor/executor_test.go [hunk 13]
```diff
@@ -2079,21 +2081,33 @@ func (s *testContextOptionSuite) TestCoprocessorPriority(c *C) {
...
+	defer func() { config.GetGlobalConfig().Log.ExpensiveThreshold = oldThreshold }()
+
+	cli.priority = pb.CommandPri_High
+	tk.MustQuery("select id from t where id = 1")
+	tk.MustQuery("select * from t1 where id = 1")
+
+	cli.priority = pb.CommandPri_Low
+	tk.MustQuery("select count(*) from t")
+	tk.MustExec("delete from t")
+	tk.MustExec("insert into t values (3)")
```
Comment (tiancaiamao) [inline executor/executor_test.go]: `insert into t values (3)` is low priority?

File: executor/executor_test.go [hunk 14]
```diff
@@ -2079,21 +2081,33 @@ func (s *testContextOptionSuite) TestCoprocessorPriority(c *C) {
...
 
 	// TODO: Those are not point get, but they should be high priority.
```
Comment (tiancaiamao) [inline executor/executor_test.go]: Have we fix this TODO now?

File: executor/compiler.go (hunk 10 above)
Comment (tiancaiamao) [reply executor/compiler.go]: ok

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**